### ISSUE TITLE
Improve ease of building (especially wrt Qt5)

### ISSUE THREAD

**Issue by CypherSignal** [devinacker/bsnes-plus#204]
On the Qt5/newdebugger branch, it would probably be worth it to do a couple things to improve the deploy and build process with regards to Qt, in two specific ways:

* Set up a Git Submodule that hosts a minimum set of Qt binary and include files, and point the makefile to that. This way, one does not have to install and configure (perhaps erroneously 🙀) their own Qt environment, and/or not require other contributors to have to juggle Qt versioning in case they're also working on other apps that have different versions of Qt
* Statically link the Qt binaries so that deployment of bsnes-plus can just be moving around a single executable file. Right now, it even needs the appropriate Qt environment installed in order to run, on Windows.

**Comment by CypherSignal** [comment]
(Note that, unless flagged otherwise, I'm willing and able to do the work on this as soon as possible, in the newdebugger branch)

**Comment by devinacker** [comment]
>Set up a Git Submodule that hosts a minimum set of Qt binary and include files, and point the makefile to that

Maybe for msys2 users, just so people don't have to settle for msys2's own horribly bloated Qt5 distribution, but I'm not going to attempt to maintain something like this for every possible build environment.

>Statically link the Qt binaries so that deployment of bsnes-plus can just be moving around a single executable file.

snesfilter and snesreader both also depend on Qt, so if you're deploying both the base application and the plugins (as I do), that's three different things in the same project all statically linking Qt, which I'm not too wild about. 

(Doing this also wouldn't be compatible with bsnes-plus' current license anyway due to the mismatched GPL versions, and that's not really something I feel like getting into at the moment)

> Right now, it even needs the appropriate Qt environment installed in order to run, on Windows.

Huh? The only parts of Qt it actually needs to run on Windows are the DLLs that would normally be included with a release (unless you're running a debug build or something in which case you already necessarily have all of that stuff available anyway).

**Comment by CypherSignal** [comment]
>Maybe for msys2 users, just so people don't have to settle for msys2's own horribly bloated Qt5 distribution, but I'm not going to attempt to maintain something like this for every possible build environment.

Hmm, after some other talking or digging may be the case that this is only required for Windows, specifically. In any case, it'd only need to be occasionally updated for one build environment.

>(Doing this also wouldn't be compatible with bsnes-plus' current license anyway due to the mismatched GPL versions, and that's not really something I feel like getting into at the moment)

Ech, yeah, fair. Out of curiosity - and effectively separate from this issue; I won't pursue this point further immediately - for prepping of binary releases, do you have a script somewhere to copy a list of needed DLLs, or had you just collected the specific ones manually?

**Comment by devinacker** [comment]
>Ech, yeah, fair. Out of curiosity - and effectively separate from this issue; I won't pursue this point further immediately - for prepping of binary releases, do you have a script somewhere to copy a list of needed DLLs, or had you just collected the specific ones manually?
Qt has its own per-platform deployment tools that bundle the needed DLLs automatically (though I tend to leave out most of the Qt5 "plugin" DLLs that bsnes doesn't actually need), though that's not currently used by the makefile or anything, so I'm just doing it manually at the moment.

**Comment by CypherSignal** [comment]
To remedy the issue with installing Qt5 for building on Windows I've got this branch, now https://github.com/devinacker/bsnes-plus/compare/newdebugger...CypherSignal:qt5-win-deploy AFAICT this reduces "required installations" down to just installing mingw - I'm able to eliminate the PATH dependency and everything builds and runs fine on my Windows machine, but I have not done a clean test to verify as such.

Note the https://github.com/CypherSignal/bsnes-plus-ext-qt submodule/repo, where the Qt code and binary dependencies lie; you should have access to that. I want to check a couple things and verify that's all A-OK (it probably is; this is similar to what, e.g. RenderDoc and Dolphin do) but in parallel want to run this by you as something that would be acceptable for a PR. Of course, the actual submodule pointed to would be transferred to you, first, for as long as you're willing to maintain bsnes+.

Basically, future updates to Qt should include updating that ext-qt repo with newer binary data from the Qt installation, and updating the submodule reference.
